acceptance-tests/strict_sni_test.go

```diff
@@ -0,0 +1,129 @@
+package acceptance_tests
+
+import (
+	"crypto/tls"
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Strict SNI", func() {
+	opsfile := `---
+# Strict SNI
+- type: replace
+  path: /instance_groups/name=haproxy/jobs/name=haproxy/properties/ha_proxy/strict_sni?
+  value: ((strict_sni))
+# Configure CA and cert chain
+- type: replace
+  path: /instance_groups/name=haproxy/jobs/name=haproxy/properties/ha_proxy/crt_list?/-
+  value:
+    snifilter:
+    - haproxy.internal
+    ssl_pem:
+      cert_chain: ((https_frontend.certificate))((https_frontend_ca.certificate))
+      private_key: ((https_frontend.private_key))
+# Declare certs
+- type: replace
+  path: /variables?/-
+  value:
+    name: https_frontend_ca
+    type: certificate
+    options:
+      is_ca: true
+      common_name: bosh
+- type: replace
+  path: /variables?/-
+  value:
+    name: https_frontend
+    type: certificate
+    options:
+      ca: https_frontend_ca
+      common_name: ((cert_common_name))
+      alternative_names: ((cert_sans))
+`
+
+	var creds struct {
+		HTTPSFrontend struct {
+			Certificate string `yaml:"certificate" json:"certificate"`
+			PrivateKey  string `yaml:"private_key" json:"private_key"`
+			CA          string `yaml:"ca" json:"ca"`
+		} `yaml:"https_frontend" json:"https_frontend"`
+		HTTPSFrontendCA struct {
+			Certificate string `yaml:"certificate" json:"certificate"`
+			PrivateKey  string `yaml:"private_key" json:"private_key"`
+		} `yaml:"https_frontend_ca" json:"https_frontend_ca"`
+	}
+
+	It("Correctly verifies the SNI", func() {
+		haproxyBackendPort := 12000
+		haproxyInfo, varsStoreReader := deployHAProxy(baseManifestVars{
+			haproxyBackendPort:    haproxyBackendPort,
+			haproxyBackendServers: []string{"127.0.0.1"},
+			deploymentName:        defaultDeploymentName,
+		}, []string{opsfile}, map[string]interface{}{
+			"cert_common_name": "haproxy.internal",
+			"cert_sans":        []string{"haproxy.internal"},
+			"strict_sni":       false,
+		}, true)
+
+		err := varsStoreReader(&creds)
+		Expect(err).NotTo(HaveOccurred())
+
+		haproxyIP := haproxyInfo.PublicIP
+
+		// Requests that use an IP rather than a hostname do not send an SNI.
+		// However the IP of HAProxy is not known until deploy time.
+		// After the first deploy, BOSH won't change the IP, so we will now
+		// redeploy and update the cert to include the IP
+		haproxyInfo, varsStoreReader = deployHAProxy(baseManifestVars{
```

Comment:
I believe there is another way to disable SNI on the client by customizing the underlying transport, so we don't have to redeploy HAproxy so often. Let's try that in a follow-up PR

Reply:
Found it:
When making the HTTP Client, in the transport we can set a TLS Config with servername:
```
		tlsConfig := &tls.Config{
			ServerName:         "put ip here",
                 }
```
This will be used as SNI if set to a string, but not be used if set to any (random) IP address.